transmat/mixins/fshash/bucket_memory.go

```diff
@@ -13,15 +13,46 @@ var _ Bucket = &MemoryBucket{}
 
 type MemoryBucket struct {
 	// my kingdom for a red-black tree or other sane sorted map implementation
-	lines []Record
+	names   []string
+	records map[string]Record
 }
 
 func (b *MemoryBucket) AddRecord(metadata fs.Metadata, contentHash []byte) {
+	if b.records == nil {
+		b.records = map[string]Record{}
+	}
+
+	name := metadata.Name.String()
+	if metadata.Type == fs.Type_Dir {
+		name += "/"
+	}
+	b.names = append(b.names, name)
+	b.records[name] = Record{name, metadata, contentHash}
+}
+
+func (b *MemoryBucket) HasRecord(metadata fs.Metadata) bool {
+	name := metadata.Name.String()
+	if metadata.Type == fs.Type_Dir {
+		name += "/"
+	}
+	_, hasRecord := b.records[name]
+	return hasRecord
+}
+
+func (b *MemoryBucket) UpdateRecord(metadata fs.Metadata, contentHash []byte) {
 	name := metadata.Name.String()
 	if metadata.Type == fs.Type_Dir {
 		name += "/"
 	}
-	b.lines = append(b.lines, Record{name, metadata, contentHash})
+	b.records[name] = Record{name, metadata, contentHash}
+}
+
+func (b *MemoryBucket) recordList() []Record {
+	var records []Record
+	for _, n := range b.names {
+		records = append(records, b.records[n])
+	}
+	return records
```

Comment:
I'm not really thrilled with this implementation because it's doing a heck of a lot of copying all at once right at the end here.

Some kind of ordered map, which could efficiently retain ordering as we insert things, and also rapidly answer the "Has" question, would almost certainly produce a better outcome.  Something perhaps like https://pkg.go.dev/github.com/google/btree or https://pkg.go.dev/github.com/petar/GoLLRB/llrb (or other things that may be coming out now as golang's generics are finally appearing in the wild).

I guess this is fine, though.  Nothing's O(n^2) or anything here.  Some memory is wasted by having a map and list as siblings, but not much.  Some garbage is generated by doing this flip into another list, but it's probably lost in the noise of all the other work that would be going on around this.  And this approach is probably the minimal diff in sheer code size.

Alright.  Fine.